### PR TITLE
Add exclude unnamed flag for push

### PR DESCRIPTION
**Pull request by emcfarlane**
This PR adds the flag `--exclude-unnamed` for `push`. On `push`, this flag will allow for unnamed modules to be present in the `buf.yaml` config. Modules are filtered by name before pushing to the remote. Local only modules, modules without a name, are excluded. All dependencies for a named module must have a valid name.

As an example the following config will work with `buf push --exclude-unnamed` as long as the module `foo` does not depend on `bar`:
```yaml
version: v2
modules:
  - path: foo
    name: buf.build/example/foo
  - path: bar
```

### CODE REVIEW

File: private/buf/cmd/buf/command/push/push.go [hunk 1]
```diff
@@ -165,6 +167,12 @@ If you set the --%s flag and/or --%s flag yourself, then the value(s) will be us
 			createDefaultLabelFlagName,
 		),
 	)
+	flagSet.BoolVar(
+		&f.IgnoredUnnamed,
+		ignoreUnnamedFlagName,
+		false,
+		"Ignore unnamed modules. Unnamed modules must not be depended on by any named modules.",
```
Comment (doriable) [inline private/buf/cmd/buf/command/push/push.go:174]: > Unnamed modules must not be depended on by any named modules.

Is this true? It seems that modules that are being pushed cannot depend on unnamed modules, but not the other way around?

Comment (emcfarlane) [reply private/buf/cmd/buf/command/push/push.go:174]: I think thats correct but wording is confusing. What about?

> Named modules must not have any unnamed dependencies.

Comment (doriable) [reply private/buf/cmd/buf/command/push/push.go:174]: Oh yes, that is a lot less confusing to me ^^"

File: private/bufpkg/bufmodule/bufmoduleapi/uploader.go [hunk 2]
```diff
@@ -107,6 +108,38 @@ func (a *uploader) Upload(
 	if err != nil {
 		return nil, err
 	}
+	// Only push named modules to the registry. Any dependencies for named modules must have a name.
+	// Local unnamed modules can be skippedl if the UploadWithSkipUnnamed option is set.
```
Comment (nicksnyder) [inline private/bufpkg/bufmodule/bufmoduleapi/uploader.go:112]:
```suggestion
	// Local unnamed modules can be skipped if the UploadWithSkipUnnamed option is set.
```

File: private/buf/cmd/buf/command/push/push.go [hunk 3]
```diff
@@ -165,6 +167,12 @@ If you set the --%s flag and/or --%s flag yourself, then the value(s) will be us
 			createDefaultLabelFlagName,
 		),
 	)
+	flagSet.BoolVar(
+		&f.SkipUnnamed,
+		skipUnnamedFlagName,
+		false,
+		"Skip unnamed modules will only push named modules. Named modules must not have any unnamed dependencies.",
```
Comment (nicksnyder) [inline private/buf/cmd/buf/command/push/push.go:174]: This doc comment is kind of confusing to parse. What about this?
```suggestion
		"Only push named modules to the BSR. Named modules must not have any unnamed dependencies.",
```

File: private/bufpkg/bufmodule/bufmoduleapi/uploader.go [hunk 4]
```diff
@@ -107,6 +108,38 @@ func (a *uploader) Upload(
...
+		if err != nil {
+			return false, err
+		}
+		if err := slicesext.Reduce(deps, func(err error, dep bufmodule.ModuleDep) error {
+			if moduleName := dep.ModuleFullName(); moduleName == nil {
+				return multierr.Append(err, fmt.Errorf("dependency %q does not have a name", dep.OpaqueID()))
+			}
+			return err
+		}, nil); err != nil {
+			return false, fmt.Errorf("All dependencies for module %s must have a name: %w.", moduleName.String(), err)
```
Comment (bufdev) [inline private/bufpkg/bufmodule/bufmoduleapi/uploader.go:132]: Using multierr here is going to result in a not-fantastic actual error message for the end-user, it'd be better if we were able to print the string `all dependencies for module %q must be named but modules %s had no name", moduleName.String(), strings.Join(allDepModuleOpaqueIDs, ", "`

File: private/buf/cmd/buf/command/push/push.go [hunk 5]
```diff
@@ -92,6 +93,7 @@ type flags struct {
 	CreateVisibility   string
 	CreateDefaultLabel string
 	SourceControlURL   string
+	SkipUnnamed        bool
```
Comment (bufdev) [inline private/buf/cmd/buf/command/push/push.go:96]: On second thought, I actually think we should call this `--exclude-unnamed` as we use `exclude` as the verb everywhere else.

Comment (emcfarlane) [reply private/buf/cmd/buf/command/push/push.go]: Renamed to exclude.

**Review by doriable — APPROVED** [review]
This looks reasonable to me :) Just putting a note here for the future, we are currently using `OpaqueID` in several places now for modules without names -- this is reasonable to me, but at some point we'll need to solidify the semantics for `OpaqueID` and adjust the docs (since the docs imply that it should not be relied upon, but we definitely do in these cases).

**Comment by bufdev** [comment]
> since the docs imply that it should not be relied upon, but we definitely do in these cases

The docs definitely should be tightened up, but just a note that they said its *structure* should not be relied upon, not that it shouldn't be relied upon as a whole. However, we should document that it can be used as a human-readable identifier of the Module.

File: private/bufpkg/bufmodule/bufmoduleapi/uploader.go (hunk 4 above)
Comment (bufdev) [reply private/bufpkg/bufmodule/bufmoduleapi/uploader.go]: This sounds right - thanks for being consistent!

**Review by bufdev — APPROVED** [review]

**Pull request merged**